add docs for buf generate include-types flag (#321)
bufbuild/docs.buf.build: docs/generate/usage.mdx

$ buf generate https://github.com/foo/bar.git --template data/generate.yaml --path proto/foo
```

### Limit to specific types

By default, `buf` builds all types in the proto file. This can be limited to

> just the types (message, enum, service) that you want to be included this image.

Author: resolved in [6a68f88](https://github.com/bufbuild/docs.buf.build/pull/321/commits/6a68f88f99243f8821c29f392d4ca4e583f31385)

> When specified, the resulting image will only include descriptors to describe

Reviewer: When what is specified? You haven't yet mentioned any config options or command-line flags. I think we need to move content around. Like move this sentence to below after introducing the config syntax and command-line flag syntax.

And then, in addition to this sentence below, you can expand on the content as to how we decide what to include. If a method is named, the request and response message types will be included. When a message is included, any type references therein (like fields that refer to other messages or enums) are also included. If any included message is extendable (proto2-only feature), its known extensions will be included. If any custom options are referenced by included elements, their definitions will also be included.

Author: resolved in [6a68f88](https://github.com/bufbuild/docs.buf.build/pull/321/commits/6a68f88f99243f8821c29f392d4ca4e583f31385)

$ buf generate https://github.com/foo/bar.git --template data/generate.yaml --path proto/foo
```

### Limit to specific types

> The `buf generate` output is an object representing your protobuf file(s)
> necessary to deserialize a protobuf message. It consists of a series of symbols,
> these may be type names (messages or enums), service names, or method
> names.

Reviewer: This isn't quite right. This is describing the output of `buf build`; the output of `buf generate` is generated source code.

Maybe I can help wordsmith a little
```suggestion
The `buf generate` output is generated code for the types defined in your Protobuf
file(s). These can be messages, enums, or even RPC services and methods (if you are
using a plugin that supports RPC frameworks).
```

> By default, `buf` builds **all** types in the proto file(s). You can request
> `buf` only return the types which you absolutely need, leaving those that
> you don't want to include in the generated output.

Reviewer: ```suggestion
By default, `buf` generates code for **all** types in your source file(s).
But you can request `buf` only generate the types which you absolutely need,
omitting everything else from the generated output.
```

This can be achieved by specifying a list of types you want to include in the
`buf.gen.yaml`. All names must be fully-qualified. If any name is unknown, the
request will fail and no output will be returned. The resulting schema contains

> these elements and all of their dependencies, which is enough information for
> the caller to construct a dynamic message for any requested types (e.g.
> requested message types or request and response types for requested service and
> method names). The resulting image will only include descriptors to describe
> the requested types.

Reviewer: I know I said this was right in Slack. But re-reading this in context, I realize it doesn't make as much sense here. While this statement is true, it is a bit confusing to mention dynamic messages when we're documenting a command that generates code (for statically known types). Also, the last sentence, about an image and descriptors, doesn't make as much sense since the result here is generated code.
```suggestion
these elements and all of their dependencies. Other types will be excluded from the
resulting generated code.
```

> If a method is named, the request and response message types will be included.
> When a message is included, any type references therein (like fields that refer
> to other messages or enums) are also included. If any included message is
> extendable (proto2-only feature), its known extensions will be included. If
> any custom options are referenced by included elements, their definitions will
> also be included.

Reviewer: I know I suggested most of this copy. But I missed an aspect of naming an individual message. Also, I think it reads better to mention dependencies of types first, and then services and methods later.
```suggestion
When a message is included, any type references therein, such as from fields that
refer to other message or enum types, are also included. If any included message
is extendable (proto2-only feature), its known extensions will be included.

If a particular method is included, its request and response message types will
also be included, but other methods that belong to the same service will not
(unless they are also included). If a service is included, all of its methods
(and their request and response types) are included.

If any included types have custom options the definitions for those custom options
will also be included.
```

> If you only want to generate code for a subset of your proto file(s), you can

Reviewer: ```suggestion
If you only want to generate code for a subset of your types, you can
```

do so via the `buf.gen.yaml`:

```diff title="buf.gen.yaml"
version: v1
+types:
+  include:
+    - buf.alpha.lint.v1.IDPaths
plugins:
  - plugin: go
    out: gen/go
    opt: paths=source_relative
  - plugin: go-grpc
    out: gen/go
    opt:
      - paths=source_relative
      - require_unimplemented_servers=false
```

> You can also use the `--include-types` flag, this usage overrides `buf.gen.yaml`

Reviewer: ```suggestion
You can also use the `--include-types` flag, which overrides whatever might
be configured in `buf.gen.yaml`.
```
